Fix Cinder support

MatMaul opened the pull request:
Fix #915

I can't test on master right now, but tested before libstorage merge.
cf codedellemc/libstorage#594

akutz commented on libstorage/api/types/types_localdevices.go:
```diff
@@ -44,7 +44,9 @@ func (l *LocalDevices) MarshalText() ([]byte, error) {
 
 	for k := range l.DeviceMap {
-		keys = append(keys, k)
+		if k != "" {
```
Hi @MatMaul,

I would swear on my life I removed the above addition recently. I remember it caused issues when removing the executor as the Marshal function removed the entries sans value. I decided it wasn't the place of a marshal function to do that. But I can find no record I actually made any changes. To be sure this file hasn't been modified in over a year!

The closest I can come is [this](https://github.com/codedellemc/libstorage/commit/2ef3d7edafe33c549fab9b8b016b5e0af66e27c0#diff-a1c38504f420f4720de9cf65512711ea):

![image](https://user-images.githubusercontent.com/101085/28987014-18fa4408-792f-11e7-853a-34c1c229df90.png)

That's in `drivers/storage/libstorage/libstorage_client_xcli.go`. It removes the devices without values in the map upon receiving the values back from the executor. Would this work for you?

MatMaul replied:
you mean... that :D ?
https://github.com/codedellemc/rexray/pull/935/commits/c1fb815490222077ed94a781da6b5860acda6e84#diff-e8770e421a1e76576af604b2c6441ad2
Either you didn't see the end of my patch or I misunderstood what you said :)
The line you are pointing there avoid empty keys but does not allow empty values, I don't remember why I put that it is probably useless.

akutz commented:
Hi @MatMaul,

Did you get a chance to review my comments?

akutz commented:
Hi @MatMaul,

Ah. It looks like mea culpa. You're dealing with the keys and I with the values. I just saw what I was attuned to see given my recent history with the same data. So if we're removing zero-length keys why update the regex to accept them when unmarshalling the same data?

> `ldRX         = regexp.MustCompile(`^(.+?)=(\S+::\S*(?::\s*,\s*\S+::\S*)*)?$`)`

Wouldn't the above pattern prevent a zero-length key or value from even being parsed? Why alter the `+` to `*` so that empty values are even allowed? I get that we have to strip data on both sides, but why allow it to be unmarshalled in the first place?

MatMaul commented:
The goal is to allow empty values and I think my changes in the regexp do that (I am bad at it so I was perhaps too broad however). You can remove mentally the first chunk of the commit dealing with empty keys and it should make sense.

MatMaul commented:
I want {"/dev/sda1": ""} possible but not {"": "uuid"}. The second part is functionally useless for my cinder use case (I think).

akutz commented:
Hi @MatMaul,

That's what I thought. Yes, your pattern is too aggressive, but that's tangential. The issue is that before the executors were removed, local devices with empty values [*were* stripped](https://github.com/codedellemc/libstorage/blob/7368564cc71be9a0dbe03ddfbaeda281232eee72/drivers/storage/libstorage/libstorage_client_xcli.go#L415-L422). When I replaced the executor CLI with code that called the executor implementations directly, the code to strip empty values was removed. This caused the framework tests via the VFS driver to fail. That's why I added the code to remove the local devices with empty values. Except it seems it's actually added in *two* locations: the one you removed and the one in `getLocalDevices`. 

Let's simplify this. We both agree on the following:

* A local device pair with an empty key should *always* be invalid.
* You want to allow empty values
* I'm saying that empty values being stripped cause the VFS tests to fail
* You removed code that strips empty values from `LocalDevices` in the XCLI source but appear to have missed the redundant code in the same file's `getLocalDevices` function.

Ultimately you want pairs with empty values, but even before the executor was removed that case wasn't allowed and there are tests that fail if empty values are allowed.

I guess my question is why the change *now*?

MatMaul commented:
I completely missed the second method in the same file, sorry for the
confusion.
We are on track now 😊
I am too tired to check now but my guess is that something changed in the
logic of a method calling LocalDevices when checking the attached devices.

Le mer. 9 août 2017 à 01:02, Schley Andrew Kutz <notifications@github.com>
a écrit :

> Hi @MatMaul <https://github.com/matmaul>,
>
> That's what I thought. Yes, your pattern is too aggressive, but that's
> tangential. The issue is that before the executors were removed, local
> devices with empty values *were* stripped
> <https://github.com/codedellemc/libstorage/blob/7368564cc71be9a0dbe03ddfbaeda281232eee72/drivers/storage/libstorage/libstorage_client_xcli.go#L415-L422>.
> When I replaced the executor CLI with code that called the executor
> implementations directly, the code to strip empty values was removed. This
> caused the framework tests via the VFS driver to fail. That's why I added
> the code to remove the local devices with empty values. Except it seems
> it's actually added in *two* locations: the one you removed and the one
> in getLocalDevices.
>
> Let's simplify this. We both agree on the following:
>
>    - A local device pair with an empty key should *always* be invalid.
>    - You want to allow empty values
>    - I'm saying that empty values being stripped cause the VFS tests to
>    fail
>    - You removed code that strips empty values from LocalDevices in the
>    XCLI source but appear to have missed the redundant code in the same file's
>    getLocalDevices function.
>
> Ultimately you want pairs with empty values, but even before the executor
> was removed that case wasn't allowed and there are tests that fail if empty
> values are allowed.
>
> I guess my question is why the change *now*?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/codedellemc/rexray/pull/935#issuecomment-321105564>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkUNe5x406sAOJx-XOmfgGsepc9GSlNks5sWOkggaJpZM4OljcS>
> .
>

akutz commented:
Hi @MatMaul,

I've rebased your PR onto `master` at https://github.com/akutz/rexray/tree/bugfix/fix-cinder. If you would care to reset this PR to my branch I will merge this PR. I also removed the second location where the local devices with empty values were being removed.

akutz commented:
Hi @MatMaul,

I also misread your RX changes. It's just fine. I am an idiot across multiple days apparently.

akutz commented:
Hi @MatMaul,

I also added a test that validates the local devices with empty values are preserved. I'm still concerned, however, that these values *were* removed prior to the executor refactor, and thus any problem that exists now should have existed then as well.

MatMaul commented:
Done. Yes I guess something changed in the logic validating the device path I'll see if I can pinpoint where it breaks.
Can you have a look at why I can't use rexray with the same CLI/conf as before since the libstorage merge ?
cf https://github.com/codedellemc/libstorage/pull/594
I am unable to test currently.

akutz commented:
Hi @MatMaul,

I finally figured out the config issue. It was resolved about a week ago, and I didn't connect it to your issue until now while triaging outstanding PRs. The issue was REX-Ray was no longer properly registering the gofig registrations and picking up defaults and registered config properties. That's fixed now if you want to rebase this off of master and give it a whirl.

We're going to try and cut RC1 today, but I've no issue including this in a later RC before release next Monday as you've shown in the past you're quite thorough and produce outstanding code!

MatMaul commented:
Rebased and tested :+1:

akutz commented:
Hi @MatMaul,

Great, thanks! Just out of curiosity, is there any reason you cannot squash your commits on this PR? I know you've wanted to keep them separate in the past for purposes of authorship, but is that applicable here?

MatMaul commented:
2 are yours and I kept 2 on my side because one is a straight revert. But I don't really care I can squash mine together, yours together, everything, let me know.

akutz commented:
Hi @cduchesne,

I don't think this will make RC1 today, but I'd like to merge it in time for RC2. The reason for my hesitance is because it occurs to me we've only tested against Cinder and VFS. I would like you to test against EBS and ScaleIO if possible to make sure the changes to the local device logic doesn't impact the other drivers. This branch is rebased enough against `master` that you shouldn't have any issues building it from source. In fact, the has the command that can be used to build [the standard binary](https://gist.github.com/anonymous/d608a617a8c984a6ffe273ef2f989a1c) as well as [other builds](https://travis-ci.org/codedellemc/rexray/builds/272558201).

benoitm76 commented:
Is this pull request fixing the wrong Volume Path bug of Cinder ?

Indeed there is a known bug where Cinder Api returning the wrong VolumePath for an attached volume in some cases : https://bugs.launchpad.net/nova/+bug/1217874

For example, Cinder API says that my volume X is attach to `/dev/vdd` but in reality on my instance it is attached on path `/dev/vdc`.

akutz commented:
Hi @benoitm76,

I'll let @MatMaul answer. I *could* tell you, but I don't want to seem like a know-it-all bigshot, and it's really Mat's baby, so he should explain. I mean, I totally *could* explain it. No lie. Definitely I could do so...but yeah, Mat should 

:)

MatMaul commented:
@akutz :) @benoitm76 I don't remember dealing explicitly with that so probably not.
A new version has been released so I'll let you check the state and we can have a look if it is still the case.
